vaadin-list-box-flow-parent/vaadin-list-box-flow/src/test/java/com/vaadin/flow/component/listbox/test/ListBoxUnitTest.java

```diff
@@ -48,6 +53,40 @@ public void setItemEnabledProvider_itemDisabled() {
         assertDisabledItem(1, false);
     }
 
+    @Test
+    public void setIdentifierProvider_shouldSelectCorrectItem_dueToIdentifier() {
+        CustomItem first = new CustomItem(1L, "First");
+        CustomItem second = new CustomItem(2L, "Second");
+        CustomItem third = new CustomItem(3L, "Third");
+        List<CustomItem> items = new ArrayList<>(Arrays.asList(first, second,
+                third));
+
+        ListBox<CustomItem> listBox = new ListBox<>();
+        ListBoxListDataView<CustomItem> listDataView = listBox.setItems(items);
+        // Setting the following Identifier Provider makes the component
+        // independent from the CustomItem's equals method implementation:
+        listDataView.setIdentifierProvider(CustomItem::getId);
+
+        listBox.setValue(new CustomItem(1L));
+
+        Assert.assertNotNull(listBox.getValue());
+        Assert.assertEquals(listBox.getValue().getName(), "First");
+
+        // Make the names similar to the name of not selected one to mess
+        // with the <equals> implementation in CustomItem:
+        first.setName("Second");
+        listDataView.refreshItem(first);
+        third.setName("Second");
+        listDataView.refreshItem(third);
+
+        // Select the item not with the reference of existing item, but instead
+        // with just the Id:
+        listBox.setValue(new CustomItem(2L));
```

Comment:
What if we set `new CustomItem(2L, "Second")` to the list box? Or even `new CustomItem("Second")`? Should we test this case also?